File: c/lc/maximum-units-on-a-truck.cc
class Solution {
public:
    static int maximumUnits(vector<vector<int>>& boxType, int truckSize) {
        sort(boxType.begin(), boxType.end(), [](const vector<int> &a, const vector<int> &b){return a[1]>b[1];});
        int sm = 0;
        int b = 0;
        while (truckSize && b < boxType.size()) {
            if (truckSize >= boxType[b][0]) {
                truckSize -= boxType[b][0];
                sm += boxType[b][0] * boxType[b][1];
                b++;
            } else {
                sm += truckSize * boxType[b][1];
                truckSize = 0;
            }
        }
        return sm;
    }
};

File: kotlin/src/dk/leetcode/MaximumUnitsOnATruck.kt
package dk.leetcode

import kotlin.math.min

class Solution2 {
    fun maximumUnits(boxTypes: Array<IntArray>, truckSize: Int): Int {
        var sz: Int = truckSize
        boxTypes.sortWith(compareBy { it[1] }) // NB!: not sortedWith()
        var tail = boxTypes.size - 1
        var sm = 0
        while (sz > 0 && tail >= 0) {
            // NB! just like in java there is no Array.pop_back()
            val boxes = boxTypes[tail][0]
            val units = boxTypes[tail][1]
            val mayTake = min(sz, boxes)
            sm += mayTake * units
            sz -= mayTake
            tail--
        }
        return sm
    }
}
